src/api/web3.ts

```diff
@@ -0,0 +1,210 @@
+    const web3 = await getProvider(options)
+
+    async function getCurrentAccount(): Promise<string> {
+        try {
+            const [acct] = await web3.eth.getAccounts()
+            return acct
+        } catch (e) {
+            throw e
+        }
+    }
+
+    async function getAccounts(): Promise<string[]> {
+        try {
+            return web3.eth.getAccounts()
+        } catch (error) {
+            console.error(new Error(error))
+            return []
+        }
+    }
+
+    const getBalance = async (account: string): Promise<string> => web3.eth.getBalance(account)
+
+    async function getBlock(block: string | number = 'latest'): Promise<Partial<Block>> {
+        try {
+            const returnedBlock: Block = await web3.eth.getBlock(block)
+            return {
+                number: returnedBlock.number,
+                timestamp: returnedBlock.timestamp,
+            }
+        } catch (e) {
+            console.log('error getting block details', e)
+            return {
+                number: 0,
+                timestamp: 0,
+            }
+        }
+    }
+
+    const getNetwork = async (): Promise<number> => web3.eth.net.getId()
+    const getNetworkName = async (id?: number | string): Promise<string> => {
+        id = id || (await getNetwork())
+
+        switch (id) {
+            case '1':
+            case 1:
+                return `mainnet`
+            case '3':
+            case 3:
+                return `ropsten`
+            case '4':
+            case 4:
+                return `rinkeby`
+            case '5':
+            case 5:
+                return `goerli`
+            default:
+                return 'private'
+        }
+    }
+
+    function getNetworkProviderURL(id: number | string): string {
+        switch (id) {
+            case '1':
+            case 1:
+                return `https://mainnet.infura.io/v3/90f210707d3c450f847659dc9a3436ea`
+            case '3':
+            case 3:
+                return `https://ropsten.infura.io/v3/90f210707d3c450f847659dc9a3436ea`
+            case '4':
+            case 4:
+                return `https://rinkeby.infura.io/v3/90f210707d3c450f847659dc9a3436ea`
+            case '5':
+            case 5:
+                return `https://goerli.infura.io/v3/90f210707d3c450f847659dc9a3436ea`
+            default:
+                return 'private'
+        }
+    }
+
+    const getTransaction = async (txHash: string): Promise<Transaction> => web3.eth.getTransaction(txHash)
+    const getTransactionReceipt = async (txHash: string): Promise<TransactionReceipt> =>
+        web3.eth.getTransactionReceipt(txHash)
+
+    /* Provider Specific */
+    const setProvider = web3.setProvider.bind(web3)
+    const resetProvider = (): void => setProvider(web3.currentProvider)
+
+    /* Provider Utils */
+    const isAddress = (address: string): boolean => web3.utils.isAddress(address)
+    const isReadOnly = (): boolean => readOnly
+
+    return {
```

Comment:
opting to keep it as a simple `function` for now, may revisit the `class` instance idea in the future

Reply:
I find the class simpler in this case, you are basically constructing an Api object with methods, it looks like a good fit. 
Also you get smaller methods instead of a big function that create everything. Anyways, I'm ok with keeping the constructor function that returns an object here, not super strong opinion on that.

But what I do think we should do at least, is to move it to its own file as I was suggesting (i.e. `Web3Api.ts`) the setup can be just importing the Api and instantiating it:

```js
async function setupWeb3(options: Web3Options): Promise<Web3Api> {
    const web3 = await getProvider(options)

    // if class based
    return new Web3API({ web3 })

    // if you want to keep using the constructor function
    return new createWeb3API({ web3 })
}
```

This way, we separate the Api implementation from the instantiation of it, and also from getting the provider logic

Reply:
Also, pls, do not mark as resolved the comments unless it's obvious that is solved. It's hard to keep track otherwise